[Bug] v0.6.0 is backwards-incompatible with v0.5.3

**Which version of MSAL Go are you using?**
Note that to get help, you need to run the latest version. 
<!-- E.g. Microsoft Authentication Library for Go 1.0.0-preview -->
github.com/AzureAD/microsoft-authentication-library-for-go@v0.6.0

**Where is the issue?**
* Public client
    * [ ] Device code flow
    * [ ] Username/Password (ROPC grant)
    * [ ] Authorization code flow 
* Confidential client
    * [ ] Authorization code flow 
    * [ ] Client credentials:
        * [ ] client secret
        * [ ] client certificate
* Token cache serialization
     * [ ] In-memory cache
* Other (please describe)

a "panic: runtime error: invalid memory address or nil pointer dereference" panic occurs at https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/a04b7709039511fc3de4594aced243090c1973a9/apps/internal/oauth/ops/accesstokens/accesstokens.go#L141

**Is this a new or an existing app?**
<!-- Ex:
a. The app is in production and I have upgraded to a new version of Microsoft Authentication Library for Go.
b. The app is in production and I haven't upgraded Microsoft Authentication Library for Go, but started seeing this issue.
c. This is a new app or an experiment.
-->
The app is in production and I have upgraded to a new version of Microsoft Authentication Library for Go.

**What version of Go are you using (`go version`)?**

<pre>
$ go version
go version go1.18.4 darwin/arm64
</pre>

**What operating system and processor architecture are you using (`go env`)?**

This occurs on both darwin/mac, and also linux.

**Repro**

```go
	testMSP := NewMSAuthProvider(nil)
	t.Run("keystoreCertError", func(t *testing.T) {
		req := &oauth2.TokenRequest{Code: code, Redirect_URI: redirectURI, ClientId: clientID, KeyStores: keystores}
		_, err := testMSP.AcquireTokenByAuthCode(context.TODO(), req)
		require.Error(t, err)
	})
```

**Expected behavior**
Calling `AcquireTokenByAuthCode` should not cause a panic (even if `Certificate.Cert` is nil)).

**Actual behavior**
Calling `AcquireTokenByAuthCode` with bad inputs causes a panic:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1010d5d4c]
```

**Possible solution**
<!--- Only if you have suggestions on a fix for the bug. -->

With v0.5.3, the func call returned an error as expected. With v0.6.0, it causes a panic.

The panic occurs when https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/a04b7709039511fc3de4594aced243090c1973a9/apps/internal/oauth/ops/accesstokens/accesstokens.go#L123 passes `nil` to the `thumbprint` func, resulting in a panic at https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/a04b7709039511fc3de4594aced243090c1973a9/apps/internal/oauth/ops/accesstokens/accesstokens.go#L141.

The root cause seems to be that PR #319 removed this guard code:

https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/c4784476597b55a050e37971dac2bb396f601938/apps/internal/oauth/ops/accesstokens/accesstokens.go#L114-L120

, which previously (in v0.5.3) prevented a Credential with a nil Cert from reaching the code which now panics in v0.6.0.

The solution would seem to be to re-add guard code which ensures that the panicing func handles bad input/state gracefully (without panicing). 

**Additional context / logs / screenshots**
Add any other context about the problem here, such as logs and screenshots.

--- a/apps/confidential/confidential.go
+++ b/apps/confidential/confidential.go
@@ -146,8 +146,23 @@ type Credential struct {
 // code requires that client.go, requests.go and confidential.go share a credential type without
 // having import recursion. That requires the type used between is in a shared package. Therefore
 // we have this.
-func (c Credential) toInternal() *accesstokens.Credential {
-	return &accesstokens.Credential{Secret: c.secret, Cert: c.cert, Key: c.key, AssertionCallback: c.assertionCallback, X5c: c.x5c}
+func (c Credential) toInternal() (*accesstokens.Credential, error) {
+	if c.secret != "" {
+		return &accesstokens.Credential{Secret: c.secret}, nil
+	}
+	if c.cert != nil {
+		if c.key == nil {
+			return nil, errors.New("missing private key for certificate")
+		}
+		return &accesstokens.Credential{Cert: c.cert, Key: c.key, X5c: c.x5c}, nil
+	}
+	if c.key != nil {
+		return nil, errors.New("missing certificate for private key")
+	}
+	if c.assertionCallback != nil {
+		return &accesstokens.Credential{AssertionCallback: c.assertionCallback}, nil
+	}
+	return nil, errors.New("invalid credential")
 }
 
 // NewCredFromSecret creates a Credential from a secret.
@@ -191,6 +206,10 @@ func NewCredFromCertChain(certs []*x509.Certificate, key crypto.PrivateKey) (Cre
 		return cred, errors.New("key must be an RSA key")
 	}
 	for _, cert := range certs {
+		if cert == nil {
+			// not returning an error here because certs may still contain a sufficient cert/key pair
+			continue
+		}
 		certKey, ok := cert.PublicKey.(*rsa.PublicKey)
 		if ok && k.E == certKey.E && k.N.Cmp(certKey.N) == 0 {
 			// We know this is the signing cert because its public key matches the given private key.
@@ -312,6 +331,11 @@ func WithAzureRegion(val string) Option {
 // will store credentials for (a Client is per user). clientID is the Azure clientID and cred is
 // the type of credential to use.
 func New(clientID string, cred Credential, options ...Option) (Client, error) {
+	internalCred, err := cred.toInternal()
+	if err != nil {
+		return Client{}, err
+	}
+
 	opts := Options{
 		Authority:  base.AuthorityPublicCloud,
 		HTTPClient: shared.DefaultClient,
@@ -329,10 +353,7 @@ func New(clientID string, cred Credential, options ...Option) (Client, error) {
 		return Client{}, err
 	}
 
-	return Client{
-		base: base,
-		cred: cred.toInternal(),
-	}, nil
+	return Client{base: base, cred: internalCred}, nil
 }
 
 // UserID is the unique user identifier this client if for.

--- a/apps/confidential/confidential_test.go
+++ b/apps/confidential/confidential_test.go
@@ -5,6 +5,7 @@ package confidential
 
 import (
 	"context"
+	"crypto"
 	"crypto/rsa"
 	"crypto/x509"
 	"encoding/base64"
@@ -239,6 +240,31 @@ func TestAcquireTokenByAuthCode(t *testing.T) {
 	}
 }
 
+func TestInvalidCredential(t *testing.T) {
+	data, err := os.ReadFile("../testdata/test-cert.pem")
+	if err != nil {
+		t.Fatal(err)
+	}
+	certs, key, err := CertFromPEM(data, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, cred := range []Credential{
+		{},
+		NewCredFromAssertionCallback(nil),
+		NewCredFromCert(nil, nil),
+		NewCredFromCert(certs[0], nil),
+		NewCredFromCert(nil, key),
+	} {
+		t.Run("", func(t *testing.T) {
+			_, err := New("client-id", cred)
+			if err == nil {
+				t.Fatal("expected an error")
+			}
+		})
+	}
+}
+
 func TestNewCredFromCertChain(t *testing.T) {
 	for _, file := range []struct {
 		path     string
@@ -353,5 +379,35 @@ func TestNewCredFromCertChain(t *testing.T) {
 				}
 			})
 		}
+	}
+}
+
+func TestNewCredFromCertChainError(t *testing.T) {
+	data, err := os.ReadFile("../testdata/test-cert.pem")
+	if err != nil {
+		t.Fatal(err)
+	}
+	certs, key, err := CertFromPEM(data, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, test := range []struct {
+		certs []*x509.Certificate
+		key   crypto.PrivateKey
+	}{
+		{nil, nil},
+		{certs, nil},
+		{nil, key},
+		{[]*x509.Certificate{}, nil},
+		{[]*x509.Certificate{}, key},
+		{[]*x509.Certificate{nil}, nil},
+		{[]*x509.Certificate{nil}, key},
+	} {
+		t.Run("", func(t *testing.T) {
+			_, err := NewCredFromCertChain(test.certs, test.key)
+			if err == nil {
+				t.Fatal("expected an error")
+			}
+		})
 	}
 }

